packages/compiler-cli/test/ngtsc/template_typecheck_spec.ts

```diff
@@ -333,52 +336,87 @@ export declare class CommonModule {
-    </div>`);
-
-      const diags = env.driveDiagnostics();
-      expect(diags.length).toBe(1);
-      expect(diags[0].messageText).toContain('does_not_exist');
-      expect(formatSpan(diags[0])).toEqual('/template.html: 1:14, 1:33');
+    describe('error locations', () => {
+      it('should be correct for direct templates', () => {
+        env.write('test.ts', `
+          import {Component, NgModule} from '@angular/core';
+      
+          @Component({
+            selector: 'test',
+            template: \`<p>
+              {{user.does_not_exist}}
+            </p>\`,
+          })
+          export class TestCmp {
+            user: {name: string}[];
+          }`);
+
+        const diags = env.driveDiagnostics();
+        expect(diags.length).toBe(1);
+        expect(diags[0].file !.fileName).toBe(_('/test.ts'));
+        expect(getCodeForDiagnostic(diags[0])).toBe('user.does_not_exist');
+      });
+
+      it('should be correct for indirect templates', () => {
+        env.write('test.ts', `
+          import {Component, NgModule} from '@angular/core';
+      
+          const TEMPLATE = \`<p>
+            {{user.does_not_exist}}
+          </p>\`;
+
+          @Component({
+            selector: 'test',
+            template: TEMPLATE,
+          })
+          export class TestCmp {
+            user: {name: string}[];
+          }`);
+
+        const diags = env.driveDiagnostics();
+        expect(diags.length).toBe(1);
+        expect(diags[0].file !.fileName).toBe(_('/test.ts') + ' (TestCmp template)');
+        expect(getCodeForDiagnostic(diags[0])).toBe('user.does_not_exist');
+        expect(getCodeForDiagnostic(diags[0].relatedInformation ![0])).toBe('TEMPLATE');
+      });
+
+      it('should be correct for external templates', () => {
+        env.write('template.html', `<p>
+          {{user.does_not_exist}}
+        </p>`);
+        env.write('test.ts', `
+          import {Component, NgModule} from '@angular/core';
+      
+
+          @Component({
+            selector: 'test',
+            templateUrl: './template.html',
+          })
+          export class TestCmp {
+            user: {name: string}[];
+          }`);
+
+        const diags = env.driveDiagnostics();
+        expect(diags.length).toBe(1);
+        expect(diags[0].file !.fileName).toBe(_('/template.html'));
+        expect(getCodeForDiagnostic(diags[0])).toBe('user.does_not_exist');
+        expect(getCodeForDiagnostic(diags[0].relatedInformation ![0])).toBe(`'./template.html'`);
+      });
     });
   });
 });
 
-function formatSpan(diagnostic: ts.Diagnostic | Diagnostic): string {
-  if (diagnostic.source !== 'angular') {
-    return '<unexpected non-angular span>';
-  }
-  const span = (diagnostic as Diagnostic).span !;
-  const fileName = span.start.file.url.replace(/^C:\//, '/');
-  return `${fileName}: ${span.start.line}:${span.start.col}, ${span.end.line}:${span.end.col}`;
+function getCodeForDiagnostic(diag: ts.Diagnostic): string {
```

Comment:
I was confused for a bit, as a diagnostic's code typically refers to its numeric code, not the source code it is produced for. How about

```suggestion
function getSourceForDiagnostic(diag: ts.Diagnostic): string {
```

Reply:
Done.